### ISSUE TITLE
merge mining tag encoding is incorrect

### ISSUE THREAD

**Issue by SWvheerden** [monero-rs/monero-rs#192]
https://github.com/monero-rs/monero-rs/blob/544a0a79332d5007663fea2d5c6ba6085852760b/src/blockdata/transaction.rs#L936C64-L936C64

It is hardcoded as 33 (32 hash + 1 x 0)
It should be 32+length of the encoded field

**Comment by Boog900** [comment]
This is also an issue in core Monero, here is a PR fixing this: https://github.com/monero-project/monero/pull/9073

I also think we should remove the Option on depth, although there are some merge merge tags on the chain without depth Monero core would fail to decode them so we should too.

**Comment by hansieodendaal** [comment]
Hi there @Boog900 and @SWvheerden, could we discuss how we want to implement this change using #202 as an example? 

_**Note:** For now I did not change the definition of `SubField::MergeMining`, just made sure that the actual encoding would be according to monero's code base, basically keeping the option in the enum but not translate it as an optional in the encoding._

My issue is that the sertialize function in monero would fail for invalid depths, and in monero-rs the norm has been that serialize should never fail with logical errors:
```c++
  bool add_mm_merkle_root_to_tx_extra(std::vector<uint8_t>& tx_extra, const crypto::hash& mm_merkle_root, size_t mm_merkle_tree_depth)
  {
    CHECK_AND_ASSERT_MES(mm_merkle_tree_depth < 32, false, "merge mining merkle tree depth should be less than 32");
    size_t start_pos = tx_extra.size();
    tx_extra.resize(tx_extra.size() + 3 + 32);
    //write tag
    tx_extra[start_pos] = TX_EXTRA_MERGE_MINING_TAG;
    //write data size
    ++start_pos;
    tx_extra[start_pos] = 33;
    //write depth varint (always one byte here)
    ++start_pos;
    tx_extra[start_pos] = mm_merkle_tree_depth;
    //write data
    ++start_pos;
    memcpy(&tx_extra[start_pos], &mm_merkle_root, 32);
    return true;
  }
```
If I add in this check
```rust
                if depth_to_encode >= VarInt(32) {
                    return Err(io::Error::new(ErrorKind::InvalidData, "Invalid merge mining depth, expected < 32"));
                }
```
Then this code will panic for invalid depths
```rust
pub fn serialize<T: Encodable + std::fmt::Debug + ?Sized>(data: &T) -> Vec<u8> {
    let mut encoder = Vec::new();
    let len = data.consensus_encode(&mut encoder).unwrap();
    debug_assert_eq!(len, encoder.len());
    encoder
}
```
example
```
called `Result::unwrap()` on an `Err` value: Custom { kind: InvalidData, error: "Invalid merge mining depth, expected < 32" }
thread 'blockdata::transaction::tests::merge_mining_sub_field_depth_32_to_255' panicked at src\consensus\encode.rs:72:51:
```

**Comment by Boog900** [comment]
I don't think we need to worry about checking that, that should be down to users who need merge mining.

Also this check got remove in https://github.com/monero-project/monero/pull/9073:

https://github.com/monero-project/monero/blob/c8214782fb2a769c57382a999eaf099691c836e7/src/cryptonote_basic/cryptonote_format_utils.cpp#L742-L765

**Comment by hansieodendaal** [comment]
@SWvheerden and @Boog900, please see https://github.com/monero-rs/monero-rs/pull/202. Just note that the code coverage report is influenced by the `merge_mining_field_upgrade_v0_18` feature flag.

**Comment by Boog900** [comment]
fixed by #202
